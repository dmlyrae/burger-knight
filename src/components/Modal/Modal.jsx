import { useEffect } from "react"
import ModalOverlay from "../ModalOverlay/ModalOverlay"
import ModalStyles from "./Modal.module.css"
import { createPortal } from "react-dom"
import {
	CloseIcon
} from "@ya.praktikum/react-developer-burger-ui-components"
import PropTypes from "prop-types"

const rootModal = document.getElementById("root-modal") 

export const Modal = function({title, closeModal, children}) {

	useEffect(() => {
		const escapeListener = function(e) {
			if (e.key === "Escape") closeModal()
		}
		document.addEventListener('keydown', escapeListener)
		return () => document.removeEventListener('keydown', escapeListener)
	}, [])

	return createPortal(
				<>
					<ModalOverlay closeModal={closeModal} />
					<div className={`${ModalStyles['modal-card']} p-10 pb-12`}>
						<div className={ModalStyles['modal-card__header']}>
							<h2 className={`${ModalStyles['card-header__title']} text text_type_main-large`}>
								{title ?? ''}
							</h2>
							<div 
								className={ModalStyles['card-header__close-button']}
							>
								<CloseIcon 
									type={"secondary"} 
									onClick={closeModal}
								/>
							</div>
						</div>
						<div className={ModalStyles['modal-card__body']}>
							{children}
						</div>
					</div>
				</>,
				rootModal
			) 
}

Modal.propTypes = {
  children: PropTypes.node,
  title: PropTypes.string,
  closeModal: PropTypes.func.isRequired
}

export default Modal