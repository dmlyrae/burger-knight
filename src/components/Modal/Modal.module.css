.modal-card {
    position: fixed;
    left: 50%;
    top: 50%;
    transform: translateX(-50%) translateY(-50%);
    max-width: 720px;
    width: 100%;
    max-height: 718px;
    background-color: #1c1c21;
    border-radius: 40px;
    margin: auto;
    opacity: 1;
    box-sizing: border-box;
    z-index: 100;
    padding-bottom: 60px;
}
.modal-card__header {
    position: relative;
    display: flex;
    justify-content: space-between;
    align-items: center;
    min-height: 45px;
}
.card-header__close-button {
    display: flex;
    align-items: center;
    justify-content: end;
    height: 60px;
    width: 60px;
    cursor: pointer;
}
.empty-header {
    position: absolute;
    right: 0;
    top: 0;
}
.modal-card__body {
    padding: 0;
    box-sizing: border-box;
}