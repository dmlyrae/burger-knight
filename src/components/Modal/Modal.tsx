import { FC, useEffect } from "react"
import ModalOverlay from "../ModalOverlay/ModalOverlay"
import ModalStyles from "./Modal.module.css"
import { createPortal } from "react-dom"
import {
	CloseIcon
} from "@ya.praktikum/react-developer-burger-ui-components"
import { TCard } from "../../types/cards"

const rootModal = document.getElementById("root-modal") as HTMLDivElement; 

interface Modal {
	title: string;
	closeModal: {
		(data:TCard): void,
		(): void,
	}
	children: React.ReactNode,
}
export const Modal:FC<Modal> = function(props):React.ReactPortal {

	const {title, closeModal, children} = props

	useEffect(() => {
		const escapeListener = function(e:KeyboardEvent) {
			if (e.key === "Escape") closeModal()
		}
		document.addEventListener('keydown', escapeListener)
		return () => document.removeEventListener('keydown', escapeListener)
	}, [])

	return createPortal(
				<>
					<ModalOverlay closeModal={closeModal} />
					<div className={`${ModalStyles['modal-card']} p-10 pb-12`}>
						<div className={ModalStyles['modal-card__header']}>
							{
								title && (
									<h2 className={`${ModalStyles['card-header__title']} text text_type_main-large`}>
										{title}
									</h2>
								) 
							}
							<div 
								className={[ModalStyles['card-header__close-button'], (title ? "": ModalStyles["empty-header"])].join(" ")}
							>
								<CloseIcon 
									type={"secondary"} 
									onClick={closeModal}
								/>
							</div>
						</div>
						<div className={ModalStyles['modal-card__body']}>
							{children}
						</div>
					</div>
				</>,
				rootModal
			) 
}

export default Modal