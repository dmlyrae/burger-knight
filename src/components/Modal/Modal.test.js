import React from 'react';
import renderer from 'react-test-renderer';
import Modal from './Modal';

it('Рендерится без ошибок', () => {
	const rootModal = document.createElement('div');
	rootModal.setAttribute("id", "root-modal");
	rootModal.style.width = "100vw";
	rootModal.style.height = "100vh";
	document.body.appendChild(rootModal);
	const tree = renderer
		.create(
			<></>
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
}); 