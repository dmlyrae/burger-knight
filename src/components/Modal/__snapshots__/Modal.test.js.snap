// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Рендерится без ошибок 1`] = `null`;
